controllers/remote/cluster_cache.go

```diff
@@ -135,21 +135,48 @@ func (t *ClusterCacheTracker) clusterAccessorExists(cluster client.ObjectKey) bo
 	return exists
 }
 
-// getClusterAccessorLH first tries to return an already-created clusterAccessor for cluster, falling back to creating a
-// new clusterAccessor if needed. Note, this method requires t.lock to already be held (LH=lock held).
-func (t *ClusterCacheTracker) getClusterAccessorLH(ctx context.Context, cluster client.ObjectKey, indexes ...Index) (*clusterAccessor, error) {
-	a := t.clusterAccessors[cluster]
+// getClusterAccessor first tries to return an already-created clusterAccessor for cluster, falling back to creating a new clusterAccessor if needed.
+func (t *ClusterCacheTracker) getClusterAccessor(ctx context.Context, cluster client.ObjectKey, indexes ...Index) (*clusterAccessor, error) {
+	log := ctrl.LoggerFrom(ctx, "cluster", cluster.Name)
+
+	loadExistingAccessor := func() *clusterAccessor {
+		t.lock.RLock()
+		defer t.lock.RUnlock()
+		return t.clusterAccessors[cluster]
+	}
+	storeAccessor := func(a *clusterAccessor) {
+		t.lock.Lock()
+		defer t.lock.Unlock()
+		t.clusterAccessors[cluster] = a
+	}
+
+	// if the accessor exists, return early
+	a := loadExistingAccessor()
 	if a != nil {
 		return a, nil
 	}
 
+	// No cluster exists, we might need to initialize one.
+	// Lock on the cluster to ensure only one accessor is initialized for the cluster.
+	unlockCluster := t.clusterLock.Lock(cluster)
+	defer unlockCluster()
+
+	// While we were waiting on the cluster lock, a different goroutine holding the lock might have initialized the accessor
+	// for this cluster successfully. If this is the case we return it.
+	a = loadExistingAccessor()
+	if a != nil {
+		return a, nil
+	}
+
+	// We are the one who needs to initialize it.
+	log.V(4).Info("creating new cluster accessor")
 	a, err := t.newClusterAccessor(ctx, cluster, indexes...)
 	if err != nil {
+		log.V(4).Info("error creating new cluster accessor")
 		return nil, errors.Wrap(err, "error creating client and cache for remote cluster")
 	}
-
-	t.clusterAccessors[cluster] = a
-
+	log.V(4).Info("storing new cluster accessor")
```

Comment:
```suggestion
	log.V(4).Info("Storing new cluster accessor")
```

Reply:
Is this log line needed?

Reply:
Strictly speaking the log lines are not needed, however they very useful for debugging. From all log lines together it is possible to figure out what happened to each attempt of creating a cluster accessor. Used these logs to verify that no accessors for the same cluster are created in parallel and that accessors for different clusters are initialized in parallel.

The logs have this high log level (V4) because they are not useful outside of debugging this code. Happy to remove them, if you think they are not providing enough value.